File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curNode = head
        prevNode = None
        bufferNode = None
        while curNode != None:

            bufferNode = curNode.next
            curNode.next = prevNode
            prevNode = curNode
            curNode = bufferNode
            # if print(f"current Node: {curNode.val}")
            # if curNode.next != None:
            #     print(f"current Node next: {curNode.next.val}")
            # else:
            #     print(f"current Node next: None")
            # if prevNode != None: 
            #     print(f"prev Node: {prevNode.val}")
            # else:
            #     print("prev Node: None")
            # if prevNode.next != None:
            #     print(f"prev Node next: {prevNode.next.val}")
            # else:
            #     print(f"prev Node next: None")

            # A -> B -> C
            # A <- B x C

        return prevNode
        

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode last = head;
        ListNode first = head;
        
        for (int i = 0; i < n+1; i++){
            if (first == null) {
                head = head.next;
                return head;
            }
            first = first.next;
        }

        while (first != null){
            first = first.next;
            last = last.next;
        }

        last.next = last.next.next;
        return head;
    }
}
